diger implementation

jasoncolburne opened the pull request:
# Rationale
`Diger` is built upon `Matter` and hashing algorithms alone. It was a pretty simple next step after implementing `Matter`.

# Testing
The code was so simple I merely exhausted the primary control flows. The digest tests ensure we don't mismatch an algorithm, they aren't intended to test the hasing functions themselves.

I added a test for 'compare' ported from KERIpy, as it exercises more code than the other tests.

# Notes
With 512-bit digests, ahead of KERIpy.

m00sey reviewed (changes requested):
We should include at least one test vector from keripy, it's been our "measure of interoperability" so far.

https://github.com/WebOfTrust/keripy/blob/5a19417a3c6e79e868ae96f55e82385de0c348d3/tests/core/test_coring.py#L4167

jasoncolburne commented:
> We should include at least one test vector from keripy, it's been our "measure of interoperability" so far.

I'll do this tonight

jasoncolburne commented:
I added the most comprehensive python test and fixed a bunch of stuff in Matter. Let me know if you want to go through anything I did. Basically, we were passing things by value and doing a lot of unnecessary clones and conversions. It broke down when the test code got more complex and needed to reuse borrowed values.

jasoncolburne commented on src/lib.rs:
```diff
@@ -1,2 +1,3 @@
+//#![deny(elided_lifetimes_in_paths)]
```
Whoops, this comment is unintentional. Let me remove that.

m00sey reviewed (approved):
lgtm

Some of those extra clones were definitely me trying to fix compilation errors around borrowing etc, you're changes make a  lot of sense so thanks for cleaning it up.
